### ISSUE TITLE
Security issue with installed executable scripts

### ISSUE THREAD

**Issue by astoff** [luarocks/luarocks#1491]
- **Platform:** Linux

- **LuaRocks version:**
  - `luarocks --version` 3.9.1

- **Description:**

Programs installed via LuaRocks can be easily tricked to load arbitrary Lua code in the current directory. Consider the following:

```
$ luarocks install luacheck
[...]
$ echo 'os.execute("echo rm -rf /")' > lfs.lua
$ luacheck --version
rm -rf /
/usr/bin/lua: /home/augusto/Projects/digestif/.luarocks/share/lua/5.4/luacheck/version.lua:21: attempt to index a boolean value (local 'lfs')
stack traceback:
	/home/augusto/Projects/digestif/.luarocks/share/lua/5.4/luacheck/version.lua:21: in main chunk
	[C]: in function 'require'
	/home/augusto/Projects/digestif/.luarocks/share/lua/5.4/luacheck/main.lua:9: in main chunk
	[C]: in function 'require'
	/home/augusto/Projects/digestif/.luarocks/lib/luarocks/rocks-5.4/luacheck/1.1.0-1/bin/luacheck:2: in main chunk
	[C]: in ?
```

As an executable script in the PATH, `luacheck` shouldn't try to load Lua files in the current directory; at least not if the user is just asking for the `--version` of a program they might not even know to be a Lua script. Of course this is not an issue with `luacheck`, but rather with the luarocks wrapper script.

As a comparison, Python also looks for "local" modules by default, but in that case the lookup is relative to the file containing the import statement, after resolving symlinks. This means that if a Python program is installed in a system directory, it can never load modules outside system directories.

Python's approach is not suitable for Lua, but the issue can be solved in different ways. I would suggest to simply remove relative directories in the wrapper script `$(which luacheck)` instead of prepending to package.path and cpath.

**Comment by daurnimator** [comment]
I guess this is saying that the wrapper that luarocks creates should explicitly set `LUA_PATH` and `LUA_CPATH`?

**Comment by astoff** [comment]
Yes. Or, to be a bit more specific, instead of prepending some stuff to `package.path` and `package.cpath`, it should completely overwrite those variables. This can be done since LuaRocks knows the locations of (the correct versions of) all dependencies of the package in question.

**Comment by daurnimator** [comment]
> This can be done since LuaRocks knows the locations of (the correct versions of) all dependencies of the package in question.

Packages themselves may be scriptable and users may rely on the package path set by the user: we shouldn't break that without a major revision.
So best can really do is prevent a shadowing attack

**Comment by astoff** [comment]
Backward compatibility issues aside, I think a scriptable package typically has its own mechanism to decide where to look for libraries. I'm only familiar with LuaTeX, and it certainly does. In any case, only the said scriptable package can decide if it's okay to load from the current directory.

>  So best can really do is prevent a shadowing attack

Please also keep in mind optional, undeclared dependencies such as
```lua
local json = require(pcall(require, "cjson") and "cjson" or "dkjson")
```

**Comment by hishamhm** [comment]
> Backward compatibility issues aside

We cannot brush backward compatibility aside. :)

> I think a scriptable package typically has its own mechanism to decide where to look for libraries. I'm only familiar with LuaTeX, and it certainly does. In any case, only the said scriptable package can decide if it's okay to load from the current directory.

Yes, only the scriptable language can decide. Ultimately, LuaRocks does not dictate how Lua applications should behave. If an application wants to sanitize its package path, or use an alternative environment variable for loading its plugins, etc., it is up to the application. Respecting the user's LUA_PATH/CPATH is the way Lua scripts behave by default. This is not even LuaRocks-specific: LuaRocks does not add `./?.lua` to the package paths anywhere — Lua does.

_Because_ LuaRocks respects the Lua defaults, if you don't want `./?.lua` to be in the package paths, you can do that today by setting the appropriate environment variables.

> So best can really do is prevent a shadowing attack

> Please also keep in mind optional, undeclared dependencies such as

LuaRocks 4 will not have optional undeclared dependencies. My plan is to bundle and preload all dependencies, and this will address the problem with the `luarocks` command raised by this issue.

Another valid issue raised here, which will also be solved by bundling, is that the `luarocks` command itself currently behaves a little different from other wrapper scripts: your example carefully uses `lfs`, a C-based dependency, because as of today the `luarocks` command only prepends the `package.path` with its own prefix, and not the `package.cpath` (because those possibly come from other places than where the Lua sources of LuaRocks was installed). Other scripts installed by LuaRocks prepend both `package.path` and `package.cpath` with the LuaRocks-paths, preventing shadowing of modules of both kinds.

**Comment by astoff** [comment]
> > Backward compatibility issues aside
>
> We cannot brush backward compatibility aside. :)

Of course not (and let me emphasize that I didn't say to _brush_ it aside).

> _Because_ LuaRocks respects the Lua defaults, if you don't want `./?.lua` to be in the package paths, you can do that today by setting the appropriate environment variables.

Yes, and I did that to the one package I have on LuaRocks. But I'm writing here as a user rather than a package author. LuaRocks should be safe by default. You can't expect every package author to take care of this, let alone every user.

To clarify a bit my point of view: if I use LuaRocks to install some program in my `/usr/local` tree, and later on I'm working on a project at `~/myproject` with its own funny `LUA_PATH` settings, then I don't want those system-level programs to be affected by the project settings. If I wanted that, I would just install a new copy of the program in question into the project-specific tree at `~/myproject`.

> Other scripts installed by LuaRocks prepend both `package.path` and `package.cpath` with the LuaRocks-paths, preventing shadowing of modules of both kinds.

This is not true (in LuaRocks 3 at least). Note that my example uses `luacheck` installed straight from LuaRocks.

**Comment by hishamhm** [comment]
> This is not true (in LuaRocks 3 at least). Note that my example uses luacheck installed straight from LuaRocks.

Ahh, missed that, you're right — in that case it's because the module is an `.so` module and you hijacked it with a `.lua` module.

> You can't expect every package author to take care of this, let alone every user.

But it _is_ up to the package author if they want to change the default behavior of Lua, or at least the system integrator, e.g. in the case of a Linux distribution setting default environment variables. If you don't want Lua programs to use the system's (or the default) `LUA_PATH`, you should probably take this to the Lua mailing list and try to convince the Lua team to change their default paths.

The LuaRocks wrappers prepend the installed rock's paths as a convenience, but it does not aim to imply any added security to a package author's program. We could just as well simply install each package's upstream scripts as-is without wrappers and tell users that setting up environment variables so that the modules from rock trees are found are up to them. After all, I don't think any other package managers try to give users this convenience... A quick look at scripts installed via CPAN, pip and RubyGems confirms this: they're just plain scripts with a default hashbang, and it's up to the system configuration to make sure modules are found in the correct order. In fact, maybe that's the way to go in the future and just drop the wrappers altogether, if our well-intentioned scripts are giving users a false sense of security.

**Comment by astoff** [comment]
Okay, you seem have a different view of the role of a package manager, at least when it comes to the installation of executable scripts. Of course, when I say LuaRocks _should_ this or that, it just means that I _think_ it should. It's one data point for you :-).

> e.g. in the case of a Linux distribution setting default environment variables.

This would be wrong. The Linux distro should not change the default behavior of the `lua` command. It should only configure _executable scripts_ to run correctly and safely.

But you are right that the issue I described with LuaRocks above applies exactly the same to Linux distros. Both should configure their executable scripts properly.

> If you don't want Lua programs to use the system's (or the default) `LUA_PATH`, you should probably take this to the Lua mailing list and try to convince the Lua team to change their default paths.

Lua could help a bit by providing a `--safe` switch to not add any relative directories from the package paths. It would be useful in shebangs and wrapper scripts. I already wrote about this to the list but the maintainers ignored the message. Maybe you have more access to them.

> A quick look at scripts installed via CPAN, pip and RubyGems confirms this

I'm not familiar with CPAN and RubyGems, but I can tell you for sure that Python does not have the problem I described here. While Python also looks for "local" modules by default, it does the lookup is relative to the input script and the interpreter location and not the current directory, so the issue I described in the original message cannot occur. A script installed at `/prefix/path/bin` can't load anything outside of `/prefix/path` by default.
